File: sparktk-core/src/main/scala/org/trustedanalytics/sparktk/frame/internal/ops/matrix/MatrixCovarianceMatrix.scala
/**
 *  Copyright (c) 2016 Intel Corporation 
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.trustedanalytics.sparktk.frame.internal.ops.matrix

import org.apache.spark.sql.Row
import org.trustedanalytics.sparktk.frame.{ DataTypes, Frame, Column }
import org.trustedanalytics.sparktk.frame.internal.{ RowWrapper, FrameState, FrameTransform, BaseFrame }
import org.apache.spark.mllib.linalg.{ DenseMatrix => DM, Matrix, Matrices }
import breeze.linalg.{ DenseMatrix => BDM, Matrix => BM, Axis, sum, DenseVector }

trait MatrixCovarianceMatrixTransform extends BaseFrame {

  def matrixCovarianceMatrix(matrixColumnName: String): Unit = {
    execute(MatrixCovarianceMatrix(matrixColumnName))
  }
}

case class MatrixCovarianceMatrix(matrixColumnName: String) extends FrameTransform {

  require(matrixColumnName != null, "Matrix column name cannot be null")

  override def work(state: FrameState): FrameState = {

    val frame = new Frame(state.rdd, state.schema)

    frame.schema.requireColumnIsType(matrixColumnName, DataTypes.matrix)
    //run the operation
    frame.addColumns(MatrixCovarianceMatrix.matrixCovarianceMatrix(matrixColumnName), Seq(Column("CovarianceMatrix_" + matrixColumnName, DataTypes.matrix)))
    FrameState(frame.rdd, frame.schema)
  }
}

object MatrixCovarianceMatrix extends Serializable {
  /**
   * Computes the covariance matrix for each matrix of the frame
   */

  def matrixCovarianceMatrix(matrixColumnName: String)(rowWrapper: RowWrapper): Row = {

    val matrix = rowWrapper.value(matrixColumnName).asInstanceOf[DM]
    val covarianceMatrix = computeCovarianceMatrix(MatrixFunctions.asBreeze(matrix))
    Row.apply(MatrixFunctions.fromBreeze(covarianceMatrix).asInstanceOf[DM])
  }

  /**
   * Computing the Covariance Matrix for a Breeze DenseMatrix
   * @param matrix Matrix whose Covariance Matrix is to be computed
   * @return Breeze DenseMatrix storing the covariance matrix
   */
  private def computeCovarianceMatrix(matrix: BDM[Double]): BDM[Double] = {
    val n = matrix.cols
    val denseMatrix: BDM[Double] = matrix.copy
    val mu: DenseVector[Double] = sum(denseMatrix, Axis._1) :* (1.0 / n)
    (0 until n).map(i => denseMatrix(::, i) :-= mu)
    val covarianceMatrix: BDM[Double] = (denseMatrix * denseMatrix.t) :* (1.0 / (n - 1))

    (covarianceMatrix + covarianceMatrix.t) :* (0.5)
  }
}

File: python/sparktk/frame/ops/matrix_covariance_matrix.py
# vim: set encoding=utf-8

#  Copyright (c) 2016 Intel Corporation 
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

def matrix_covariance_matrix(self, matrix_column_name):

    """

    Compute the Covariance Matrix of matrices stored in a frame

    Parameters
    ----------

    :param matrix_column_name: Name of the column to compute the covariance matrix on
    :return: (Frame) returns the frame with a new column storing the covariance matrix for the corresponding matrix

    Calculate the covariance matrix for each matrix in column 'matrix_column_name' of a frame using the following:

    Element (i,j) of the covariance matrix for a given matrix X is computed as: ((Xi - Mi)(Xj - Mj))
    where Mi is the mean

    Examples
    --------
        >>> from sparktk import dtypes
        >>> data = [[1, [[1,2,3,5],[2,3,5,6],[4,6,7,3],[8,9,2,4]]]]
        >>> schema = [('id', int),('pixeldata', dtypes.matrix)]
        >>> my_frame = tc.frame.create(data, schema)

        >>> my_frame.inspect()
        [#]  id  pixeldata
        ============================
        [0]   1  [[ 1.  2.  3.  5.]
        [ 2.  3.  5.  6.]
        [ 4.  6.  7.  3.]
        [ 8.  9.  2.  4.]]


        Compute the covariance matrix for the matrices in 'pixeldata' column of the frame
        >>> my_frame.matrix_covariance_matrix('pixeldata')

        A new column gets added to the existing frame storing the covariance matrix
        >>> my_frame.inspect()
        [#]  id  pixeldata
        ============================
        [0]   1  [[ 1.  2.  3.  5.]
        [ 2.  3.  5.  6.]
        [ 4.  6.  7.  3.]
        [ 8.  9.  2.  4.]]
        <BLANKLINE>
        [#]  CovarianceMatrix_pixeldata
        ============================================================
        [0]  [[  2.91666667   3.          -1.          -3.75      ]
        [  3.           3.33333333  -0.33333333  -5.        ]
        [ -1.          -0.33333333   3.33333333  -1.        ]
        [ -3.75        -5.          -1.          10.91666667]]

    """

    self._scala.matrixCovarianceMatrix(matrix_column_name)